backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,158 @@
+
+    const attemptSummary = await attemptSummaryPromise;
+    const durationSummary = await durationSummaryPromise;
+    const peerSummary = await peerSummaryPromise;
+    const heatmapData = await heatmapDataPromise;
+    const networkData = await networkDataPromise;
+
+    return {
+      attemptSummary,
+      durationSummary,
+      peerSummary,
+      heatmapData,
+      networkData,
+    };
+  }
+
+  private async getAttemptSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    const map = {
+      [Difficulty.EASY]: 0,
+      [Difficulty.MEDIUM]: 0,
+      [Difficulty.HARD]: 0,
+    };
+    for (const session of data) {
+      const { difficulty } = session.question;
+      map[difficulty] = map[difficulty] + 1;
+    }
+    return map;
+  }
+
+  private async getDurationSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subDays(new Date(), 30),
+        },
+      },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    return data
+      .filter((session) => !!session.endTime)
+      .map((session) => {
+        return {
+          difficulty: session.question.difficulty as Difficulty,
+          timetaken: differenceInSeconds(
+            session.endTime || new Date(), // Should be defined by selection.
+            session.startTime,
+          ),
+          date: session.startTime,
+        };
+      });
+  }
+
+  private async getPeerSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { title: true },
+        },
+        users: true,
+      },
+    });
+    return data.map((session) => {
+      return {
+        userName: session.users.find((user) => user.id !== userId)?.name || "",
+        questionTitle: session.question.title,
+        date: session.startTime,
+      };
+    });
+  }
+
+  private async getHeatmapDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+    });
+    return data.map((session) => {
+      return { date: session.startTime };
+    });
+  }
+
+  private async getNetworkDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
```

Comment:
I think there is an implicit assumption here that the user has to complete the topic rather than attempt the topic. Should i change the frontend title to "Topics Completed" instead?

Reply:
Technically this should be at most an off by one error since only the current session that the user is in will not be counted. I think both "Attempted" and "Completed" are more or less the same here.